NPE in WebSocket client subscriptions

caught suddenly while debugging https://github.com/nspcc-dev/neofs-node/pull/2444, don’t have a special context. I subscribe to new blocks and notary requests
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x7bc119]

goroutine 14 [running]:
github.com/nspcc-dev/neo-go/pkg/rpcclient.(*WSClient).notifySubscribers(0xc000224480, {0x16?, {0xa11180?, 0xc000283ea0?}})
	/home/ll/go/pkg/mod/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/wsclient.go:599 +0x299
github.com/nspcc-dev/neo-go/pkg/rpcclient.(*WSClient).wsReader(0xc000224480)
	/home/ll/go/pkg/mod/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/wsclient.go:513 +0x545
created by github.com/nspcc-dev/neo-go/pkg/rpcclient.NewWS
	/home/ll/go/pkg/mod/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/wsclient.go:422 +0x638
```

may be i violated some strict requirement on my side, but didn't find it after a quick look

--- a/pkg/rpcclient/wsclient.go
+++ b/pkg/rpcclient/wsclient.go
@@ -960,6 +960,14 @@ func (c *WSClient) UnsubscribeAll() error {
 // after WS RPC unsubscription request is completed. Until then the subscriber channel
 // may still receive WS notifications.
 func (c *WSClient) performUnsubscription(id string) error {
+	c.subscriptionsLock.RLock()
+	rcvrWas, ok := c.subscriptions[id]
+	c.subscriptionsLock.RUnlock()
+
+	if !ok {
+		return errors.New("no subscription with this ID")
+	}
+
 	var resp bool
 	if err := c.performRequest("unsubscribe", []any{id}, &resp); err != nil {
 		return err
@@ -975,6 +983,15 @@ func (c *WSClient) performUnsubscription(id string) error {
 	if !ok {
 		return errors.New("no subscription with this ID")
 	}
+
+	cleanUpSubscriptions := true
+	if rcvrWas.Receiver() != rcvr.Receiver() {
+		// concurrent subscription has been done and been overwritten; this
+		// is not this routine's subscription, cleanup only receivers map
+		rcvr = rcvrWas
+		cleanUpSubscriptions = false
+	}
+
 	ch := rcvr.Receiver()
 	ids := c.receivers[ch]
 	for i, rcvrID := range ids {
@@ -988,7 +1005,9 @@ func (c *WSClient) performUnsubscription(id string) error {
 	} else {
 		c.receivers[ch] = ids
 	}
-	delete(c.subscriptions, id)
+	if cleanUpSubscriptions {
+		delete(c.subscriptions, id)
+	}
 	return nil
 }
 

